src/grammar.rs

```diff
@@ -423,54 +423,57 @@ impl Grammar {
     }
 
     /// Determine if the starting rule of the grammar can take us to a terminal state, i.e. all
-    /// Term::Terminal types. Used as a check before generation, to tell if we can safely attempt
+    /// [`Term::Terminal`] types. Used as a check before generation, to tell if we can safely attempt
     /// to generate sentences without risking infinite loops
     pub(crate) fn terminates(&self) -> bool {
-        if self.starting_term().is_none() {
+        let Some(starting_term) = self.starting_term() else {
             // if there are no rules, there's nothing to do so... it terminates
             return true;
-        }
-
-        let starting_term: &Term = self.starting_term().unwrap();
+        };
 
-        let mut terminating_rules: Vec<Term> = vec![];
+        let mut terminating_rules: Vec<&Term> = vec![];
 
         // collect 'every non-terminal that produces a sequences of terminals'
         for p in self.productions.iter() {
             if p.has_terminating_expression(None) {
-                terminating_rules.push(p.lhs.clone())
+                terminating_rules.push(&p.lhs);
             }
         }
 
         if terminating_rules.is_empty() {
             return false;
         }
 
-        // 'recursively mark every non-terminal that can produce a sequence of marked elements'
-        terminating_rules = self.collect_terminating_rules(&terminating_rules);
+        let mut is_progress = true;
 
-        // not strictly necessary
-        terminating_rules.sort();
-        terminating_rules.dedup();
+        while is_progress {
```

Comment:
Instead of this `is_progress` loop, a work queue of only the affected production rules could be more efficient. But locally, my tests showed this performance advantage is not meaningful until large grammars (>100 productions). So for now, I think simply re-checking all productions is good enough